addons/account/tests/test_account_account.py

```diff
@@ -623,6 +623,71 @@ def test_generate_account_suggestions(self):
+            account_query = account._as_query()
+            placeholder_code_sql = account_query.select(account._field_to_sql('account_account', 'placeholder_code', account_query))
+            placeholder_code = self.env.execute_query(placeholder_code_sql)[0][0]
+            return placeholder_code
+
+        branch = self.env['res.company'].create({
+            'name': 'Branch',
+            'parent_id': self.company_data['company'].id,
+        })
+        account = self.env['account.account'].create([{
+            'name': 'My account',
+            'company_ids': [Command.set(self.company_data['company'].ids)],
+            'code_mapping_ids': [
+                Command.create({'company_id': self.company_data['company'].id, 'code': '180001'}),
+            ]
+        }])
+
+        # Use `with_company` in order to force an order for `allowed_company_ids` in the context.
+        self.assertEqual(account.with_company(self.company_data['company']).placeholder_code, '180001')
+        self.assertEqual(get_placeholder_code_via_sql(account.with_company(self.company_data['company'])), '180001')
+
+        self.assertEqual(account.with_company(branch).placeholder_code, '180001')
+        self.assertEqual(get_placeholder_code_via_sql(account.with_company(branch)), '180001')
+
+        account_context_1 = account.with_context({
+            'allowed_company_ids': [self.company_data_2['company'].id, self.company_data['company'].id]
+        })
+        self.assertEqual(account_context_1.placeholder_code, '180001 (company_1_data)')
+        self.assertEqual(get_placeholder_code_via_sql(account_context_1), '180001 (company_1_data)')
+
+        account_context_2 = account.with_context({
+            'allowed_company_ids': [self.company_data_2['company'].id, branch.id]
+        })
+        self.assertEqual(account_context_2.placeholder_code, False)
+        self.assertEqual(get_placeholder_code_via_sql(account_context_2), None)
+
+    def test_account_returned_by_search_if_sudo(self):
+        """ Test that even if an account isn't accessible by the current user, it is returned by a search in sudo mode. """
+        account = self.env['account.account'].with_company(self.company_data_2['company']).create([{
+            'name': 'Account in Company 2',
+            'code': '180002',
+        }])
+
+        # This user can't access company 2, so it can't access the created account.
+        user_that_cannot_access_company_2 = new_test_user(
+            self.env,
+            name="User that can't access company 2",
+            login='user_that_cannot_access_company_2',
+            password='user_that_cannot_access_company_2',
+            email='user_that_cannot_access_company_2@test.com',
+            groups_id=self.get_default_groups().ids,
+            company_id=self.env.company.id,
+        )
+
+        searched_account = self.env['account.account'].with_user(user_that_cannot_access_company_2).sudo().search([('id', '=', account.id)])
```

Comment:
What does this have to do with `placeholder_code`?

Reply:
This search is calling `_field_to_sql` for `placeholder_code` to generate the ORDER BY (because `placeholder_code` is in the model's `_order`. 
The use of JOIN rather than LEFT JOIN meant that the `placeholder_code` _field_to_sql actually filtered out some of the results from the search.